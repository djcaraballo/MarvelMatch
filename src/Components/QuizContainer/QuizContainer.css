@import url('https://fonts.googleapis.com/css?family=Fjalla+One|Orbitron');

.answer-options {
  margin: 2em auto 10em;
  padding: 0;
}

