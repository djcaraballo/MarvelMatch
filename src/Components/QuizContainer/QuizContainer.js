import React from 'react'
import PropTypes from 'prop-types'
import Question from '../Questions/Questions'
import QuestionCount from '../QuestionCount/QuestionCount'
import AnswerOption from '../AnswerOption/AnswerOption'

const QuizContainer = (props) => {
  const {
    answer,
    answerOptions,
    counter,
    question,
    questionId,
    questionTotal,
    onAnswerSelected,
  } = props
  
  const renderAnswerOptions = (key) => {
    return (
      <AnswerOption
          key={key.content}
          answerContent={key.content}
          answerType={key.type}
          answer={answer}
          questionId={questionId}
          onAnswerSelected={onAnswerSelected}
      />
    )
  }

  return (
    <div className='quiz-container'>
      <QuestionCount
        counter={questionId}
        total={questionTotal}
      />
      <Question text={question} />
      <ul className='answer-options'>
        {answerOptions.map(renderAnswerOptions)}
      </ul>
    </div>
  )
}

QuizContainer.propTypes = {
  answer: PropTypes.string,
  answerOptions: PropTypes.array,
  counter: PropTypes.number,
  question: PropTypes.string,
  questionId: PropTypes.number,
  questionTotal: PropTypes.number,
  onAnswerSelected: PropTypes.func
}

export default QuizContainer


