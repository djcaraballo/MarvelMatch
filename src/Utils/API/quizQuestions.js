const quizQuestions = [
  {
    question: "What color are your eyes?",
    answers: [
      {
        type: "blue",
        content: "Blue"
      },
      {
        type: "brown",
        content: "Brown"
      },
      {
        type: "green",
        content: "Green"
      },
      {
        type: "other",
        content: "Other"
      },
    ]
  },
  {
    question: "What color is your hair?",
    answers: [
      {
        type: "blond",
        content: "Blonde"
      },
      {
        type: "black",
        content: "Black"
      },
      {
        type: "red",
        content: "Red"
      },
      {
        type: "auburn",
        content: "Brown"
      },
      {
        type: "alternate",
        content: "I don't even know anymore!"
      }
    ]
  },
  {
    question: "How tall are you?",
    answers: [
      {
        type: "tall",
        content: "I'm pretty tall"
      },
      {
        type: "short",
        content: "I'm kinda short"
      },
      {
        type: "average",
        content: "I think I'm average"
      },
    ]
  },
  {
    question: "With which race do you most identify?",
    answers: [
      {
        type: "human",
        content: "I'm human!"
      },
      {
        type: "alien",
        content: "I'm alien!"
      },
      {
        type: "spirit",
        content: "I've transcended physical form!"
      },
      {
        type: "mutant",
        content: "I'm a mutant!"
      },
      {
        type: "cyborg",
        content: "I'm a cyborg!"
      }
    ]
  }
];

export default quizQuestions;

