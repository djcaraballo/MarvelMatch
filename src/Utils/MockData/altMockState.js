export const altMockState = {
      counter: 0,
      questionId: 1,
      question: "What color are your eyes?",
      answerOptions: [
        {
          type: "blue",
          content: "Blue"
        },
        {
          type: "brown",
          content: "Brown"
        },
        {
          type: "green",
          content: "Green"
        },
        {
          type: "other",
          content: "Other"
        },
      ],
      answer: '',
      answersCount: {
        blue: 1,
        brown: 0,
        green: 0,
        other: 0,
        blond: 0,
        black: 1,
        red: 0,
        auburn: 0,
        alternate: 0,
        human: 0,
        alien: 1,
        spirit: 0,
        mutant: 0,
        cyborg: 0
      },
      result: ''
    }

