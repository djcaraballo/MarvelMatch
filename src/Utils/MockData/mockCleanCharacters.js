export const cleanCharacters = [
  {
    name: "A-Bomb (HAS)", 
    id: 1017100, 
    description: "Rick Jones has been Hulk's best bud since day one,…ses it like a giant bowling ball of destruction! ", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/76/a-bomb?utm_campaig…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "A.I.M.", 
    id: 1009144, 
    description: "AIM is a terrorist organization bent on destroying the world.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/6/20/52602f21f29ec/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/77/aim.?utm_campaign=apiRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Abomination (Emil Blonsky)", 
    id: 1009146, 
    description: "Formerly known as Emil Blonsky, a spy of Soviet Yu…ransformed Bruce Banner into the incredible Hulk.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/9/50/4ce18691cbf04/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/81/abomination?utm_ca…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Adam Warlock", 
    id: 1010354, description: "Adam Warlock is an artificially created human who …ocoon at a scientific complex called The Beehive.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/a/f0/5202887448860/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/comics/characters/1010354/adam_w…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Agent X (Nijo)", 
    id: 1011031, 
    description: "Originally a partner of the mind-altering assassin…inds crime family, which included Nijo's brother.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/101/agent_x?utm_campa…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Alex Wilder", 
    id: 1010755, 
    description: "Despite being the only one of the Runaways without…ect, as well as prodigy-level logic and strategy.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/2/c0/4c00377144d5a/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/2820/alex_wilder?utm_…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Amun", 
    id: 1010905, 
    description: "Amun is a ruthless teenage assassin, employed by t…the ritual to initiate the Wasp's new chosen one.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/140/amun?utm_campaign…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Arnim Zola", 
    id: 1009740, 
    description: "The frail, dwarfish Arnim Zola was born in 1930s S… world's leading biochemist and genetic engineer.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/8/b0/4c00393a4cb7c/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/2790/arnim_zola?utm_c…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Avengers", 
    id: 1009165, 
    description: "Earth's Mightiest Heroes joined forces to take on …s have come to be regarded as Earth's No. 1 team.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/9/20/5102c774ebae7/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/68/avengers?utm_campa…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
  {
    name: "Azazel (Mutant)", 
    id: 1011766, 
    description: "A mutant from biblical times, Azazel is the ruler …at the Earth and everything on it belongs to him.", 
    imageURL: "http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/standard_fantastic.jpg", 
    siteURL: "http://marvel.com/characters/227/azazel?utm_campai…piRef&utm_source=a4d95911b60999871aaed8a8c43823f6"
  }
]
