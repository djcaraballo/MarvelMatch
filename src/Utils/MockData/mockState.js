export const mockState = {
      counter: 0,
      questionId: 1,
      question: "What color are your eyes?",
      answerOptions: [
        {
          type: "blue",
          content: "Blue"
        },
        {
          type: "brown",
          content: "Brown"
        },
        {
          type: "green",
          content: "Green"
        },
        {
          type: "other",
          content: "Other"
        },
      ],
      answer: '',
      answersCount: {
        blue: 0,
        brown: 0,
        green: 0,
        other: 0,
        blond: 0,
        black: 0,
        red: 0,
        auburn: 0,
        alternate: 0,
        human: 0,
        alien: 0,
        spirit: 0,
        mutant: 0,
        cyborg: 0
      },
      result: ''
    }

