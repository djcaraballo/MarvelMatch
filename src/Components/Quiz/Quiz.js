import React, { Component } from 'react'
import PropTypes from 'prop-types'
import QuizContainer from '../QuizContainer/QuizContainer'
import Question from '../Questions/Questions'
import quizQuestions from '../../Utils/API/quizQuestions'
import Result from '../Result/Result'

class Quiz extends Component {
  constructor (props) {
    super(props)
    this.state = {
      counter: 0,
      questionId: 1,
      question: '',
      answerOptions: [],
      answer: '',
      answersCount: {
        blueEyedHeroes: 0,
        brownEyedHeroes: 0,
        greenEyedHeroes: 0,
        otherColoredEyedHeroes: 0,
        blondeHairedHeroes: 0,
        blackHairedHeroes: 0,
        redHairedHeroes: 0,
        brownHairedHeroes: 0,
        otherHairedHeroes: 0,
        tallHeroes: 0,
        shortHeroes: 0,
        averageHeroes: 0,
      },
      result: ''
      //change to object once everything is working 
    }
  }

  componentWillMount() {
    this.setState({
      question: quizQuestions[0].question,
      answerOptions: quizQuestions[0].answers
    })
  }

  handleAnswerSelected = (e) => {
    this.setUserAnswer(e.currentTarget.value)
    if(this.state.questionId < quizQuestions.length) {
      setTimeout(() => this.setNextQuestion(), 300)
    } else {
      setTimeout(() => this.setResults(this.getResults()), 300)
    }
  }

  setUserAnswer(answer) {
    this.setState(state => 
      ({
      answersCount: {
        ...state.answersCount,
        [answer]: state.answersCount[answer]++
      },
      answer: answer
    }))
  }

  setNextQuestion() {
    const counter = this.state.counter + 1
    const questionId = this.state.questionId + 1
    this.setState({
      counter: counter,
      questionId: questionId,
      question: quizQuestions[counter].question,
      answerOptions: quizQuestions[counter].answers,
      answer: '',
    })
  }

  getResults() {
    const answersCount = this.state.answersCount
    const answersCountKeys = Object.keys(answersCount)
    const answersCountValues = answersCountKeys.map((key) => answersCount[key])
    const maxAnswerCount = Math.max.apply(null, answersCountValues)
    return answersCountKeys.filter((key) => answersCount[key] === maxAnswerCount)
  }

  setResults(result) {
    if (result.length === 1) {
      this.setState({ result: result[0] })
    } else {
      this.setState({ result: 'Undetermined' })
    }
  }

  renderQuizContainer() {
    const { answer, answerOptions, questionId, question } = this.state
    return (
      <QuizContainer 
        answer={answer}
        answerOptions={answerOptions}
        questionId={questionId}
        question={question}
        questionTotal={quizQuestions.length}
        onAnswerSelected={this.handleAnswerSelected}
      />
    )
  }

  renderResult() {
    return (
      <Result quizResult={this.state.result} />
    )
  }

  render () {
    return (
      <div className="quiz-wrapper" key={this.state.questionId}>
        <h1>Take the Quiz and match a Marvel Character!</h1>
        {this.state.result ? this.renderResult() : this.renderQuizContainer()}
      </div>
    )
  }
}

export default Quiz
