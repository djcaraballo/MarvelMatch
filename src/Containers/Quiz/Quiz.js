import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import QuizContainer from '../../Components/QuizContainer/QuizContainer'
import quizQuestions from '../../Utils/API/quizQuestions'
import Result from '../../Components/Result/Result'

export class Quiz extends Component {
  constructor (props) {
    super(props)
    this.state = {
      counter: 0,
      questionId: 1,
      question: '',
      answerOptions: [],
      answer: '',
      answersCount: {
        blue: 0,
        brown: 0,
        green: 0,
        other: 0,
        blond: 0,
        black: 0,
        red: 0,
        auburn: 0,
        alternate: 0,
        human: 0,
        alien: 0,
        spirit: 0,
        mutant: 0,
        cyborg: 0
      },
      result: ''
    }
  }

  componentWillMount() {
    this.setState({
      question: quizQuestions[0].question,
      answerOptions: quizQuestions[0].answers
    })
  }

  handleAnswerSelected = (e) => {
    this.setUserAnswer(e.target.value)
    if(this.state.questionId < quizQuestions.length) {
      setTimeout(() => this.setNextQuestion(), 400)
    } else {
      setTimeout(() => this.setResults(this.getResults()), 400)
    }
  }

  setUserAnswer(answer) {
    this.setState({
      answersCount: {
        ...this.state.answersCount, 
        [answer]: this.state.answersCount[answer]+=1
      },
      answer: answer
    })
  }

  setNextQuestion() {
    const counter = this.state.counter + 1
    const questionId = this.state.questionId + 1
    this.setState({
      counter: counter,
      questionId: questionId,
      question: quizQuestions[counter].question,
      answerOptions: quizQuestions[counter].answers,
      answer: '',
    })
  }

  getResults() {
    const answersCount = this.state.answersCount
    const answersCountKeys = Object.keys(answersCount)
    const answersCountValues = answersCountKeys.map(key => answersCount[key])
    const possibleResults = this.props.characters.filter((character) => {
      answersCountKeys.forEach((key) => {
        if(character.appearance.hairColor && key === character.appearance.hairColor.toLowerCase() && answersCount[key] !== 0) {
          character.matchTally++
        } else if(character.appearance.eyeColor && key === character.appearance.eyeColor.toLowerCase() && answersCount[key] !== 0) {
          character.matchTally++
        } else if(character.appearance.race && key === character.appearance.race.toLowerCase() && answersCount[key] !== 0) {
          character.matchTally++
        }
      })
      return character.matchTally !== 0
    })
    return possibleResults
  }

  setResults(result) {
    const index = Math.floor(Math.random() * Math.floor(result.length))
    if (result.length >= 1) {
      this.setState({ result: result[index].name })
    } else {
      this.setState({ result: 'Undetermined' })
    }
  }

  handleCharacterMatch = () => {
    const charMatch = this.props.characters.find((character) => {
      return character.name === this.state.result
    })
    return charMatch
  }

  renderQuizContainer() {
    const { answer, answerOptions, questionId, question } = this.state
    return (
      <QuizContainer 
        answer={answer}
        answerOptions={answerOptions}
        questionId={questionId}
        question={question}
        questionTotal={quizQuestions.length}
        onAnswerSelected={this.handleAnswerSelected}
      />
    )
  }

  renderResult() {
    const charMatch = this.handleCharacterMatch()
    return (
      <Result 
        quizResult={this.state.result}
        characterMatch={charMatch} 
      />
    )
  }

  render () {
    return (
      <div className="quiz-wrapper" key={this.state.questionId}>
        <h1>Take the Quiz and match a Marvel Character!</h1>
        {this.state.result ? this.renderResult() : this.renderQuizContainer()}
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  characters: state.characters
})

Quiz.propTypes = {
  characters: PropTypes.array
}

export default connect(mapStateToProps, null)(Quiz)
