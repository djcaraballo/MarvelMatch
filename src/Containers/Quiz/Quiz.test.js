import React from 'react'
import {Quiz, mapStateToProps } from './Quiz'
import { shallow } from 'enzyme'
import { cleanCharacters } from '../../Utils/MockData/mockCleanCharacters'
import { mockState } from '../../Utils/MockData/mockState'
import quizQuestions from '../../Utils/API/quizQuestions'
import QuizContainer from '../../Components/QuizContainer/QuizContainer'

let wrapper

beforeEach(() => {
  wrapper = shallow(<Quiz characters={cleanCharacters} />)
})

describe('Quiz Component', () => {
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should start with the correct default state', () => {
    expect(wrapper.state()).toEqual(mockState)
  })

  describe('setUserAnswer function', () => {
    it('should set state with the correct values for answersCount and answer', () => {
      const mockAnswer = 'Something'
      const expectedAnswerCount = {...wrapper.state().answersCount, Something: NaN}
      wrapper.instance().setUserAnswer(mockAnswer)
      expect(wrapper.state().answersCount).toEqual(expectedAnswerCount)
      expect(wrapper.state().answer).toEqual(mockAnswer)
    })
  })

  describe('setNextQuestion function', () => {
    it('should set state with the correct values', () => {
      const counter = 1
      const questionId = 2
      wrapper.instance().setNextQuestion()
      expect(wrapper.state().counter).toEqual(1)
      expect(wrapper.state().questionId).toEqual(2)
      expect(wrapper.state().question).toEqual(quizQuestions[counter].question)
      expect(wrapper.state().answerOptions).toEqual(quizQuestions[counter].answers)
    })
  })

  describe('setResults function', () => {
    it('should set state with the correct result if there is a result', () => {
      const mockResult = [{name: 'Some'}, {name: 'Things'}]
      const index = 0
      wrapper.instance().setResults(mockResult)
      expect(wrapper.state().result).toEqual('Some')
    })

    it('should set state with Undetermined if there is no result', () => {
      const mockResult = []
      const index = 1
      wrapper.instance().setResults(mockResult)
      expect(wrapper.state().result).toEqual('Undetermined')
    })
  })

  describe('handleCharacterMatch function', () => {
    it('should return a character object with a matching name', () => {
      wrapper.state().result = "A-Bomb (HAS)"
      const character = wrapper.instance().handleCharacterMatch()
      expect(character.name).toEqual("A-Bomb (HAS)")
    })
  })

  describe('renderQuizContainer function', () => {
    it('should return an instance of the QuizContainer component', () => {
      const handleAnswerSelected = jest.fn()
      const expected = 
        <QuizContainer 
          answer={mockState.answer}
          answerOptions={mockState.answerOptions}
          questionId={mockState.questionId}
          question={mockState.question}
          questionTotal={quizQuestions.length}
          onAnswerSelected={wrapper.instance().handleAnswerSelected}
        />
      const value = wrapper.instance().renderQuizContainer()
      expect(value).toEqual(expected)
    })
  })

  describe('renderResult function', () => {
    xit('should return an instance of the Result component', () => {

    })
  })
})