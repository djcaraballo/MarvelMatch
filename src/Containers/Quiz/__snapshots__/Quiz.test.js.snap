// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Component should match the snapshot 1`] = `
<div
  className="quiz-wrapper"
  key="1"
>
  <div>
    <h1>
      Take the Quiz and match a Marvel Character!
    </h1>
    <QuizContainer
      answer=""
      answerOptions={
        Array [
          Object {
            "content": "Blue",
            "type": "blue",
          },
          Object {
            "content": "Brown",
            "type": "brown",
          },
          Object {
            "content": "Green",
            "type": "green",
          },
          Object {
            "content": "Other",
            "type": "other",
          },
        ]
      }
      onAnswerSelected={[Function]}
      question="What color are your eyes?"
      questionId={1}
      questionTotal={4}
    />
  </div>
</div>
`;
